### PR TITLE
Speed up copyFonts

### PR DESCRIPTION
**Pull request by apeschar**
### Description

When a lot of fonts are used, the `copyFonts` step can take multiple seconds. It currently deletes and again copies the fonts from the Nix store, even if the files are identical.

This PR makes two changes:

1. Skip font installation if the derivation hasn't changed: the path to the build in the Nix store is written to a file. As long as the Nix store path doesn't change (ie, the font build is the same), the copyFonts step is skipped.
1. Use `rsync` instead of `install` to copy font files. Rsync will only copy files that are new or have actually changed.

With these changes the `copyFonts` step is instant when no fonts are changed, and quite a lot faster when there are changes.

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [ ] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

### CODE REVIEW

File: modules/targets/darwin/fonts.nix [hunk 1]
```diff
@@ -10,18 +10,24 @@ let
...
   fonts = "${fontsEnv}/share/fonts";
+  installDir = "${homeDir}/Library/Fonts/HomeManager";
+  versionFile = "${installDir}/derivation";
 in {
   # macOS won't recognize symlinked fonts
   config = mkIf pkgs.stdenv.hostPlatform.isDarwin {
     home.activation.copyFonts = hm.dag.entryAfter [ "writeBoundary" ] ''
       copyFonts() {
-        rm -rf ${homeDir}/Library/Fonts/HomeManager || :
+        if [[ $(cat ${versionFile} 2>/dev/null) = ${fontsEnv} ]]; then
```
Comment (berbiche) [inline modules/targets/darwin/fonts.nix:20]:
```suggestion
        if [[ $(cat ${versionFile} 2>/dev/null) = ${escapeShellArg fontsEnv} ]]; then
```

File: modules/targets/darwin/fonts.nix [hunk 2]
```diff
@@ -10,18 +10,24 @@ let
...
-        find -L "${fonts}" -type f -printf '%P\0' | while IFS= read -rd "" f; do
-          $DRY_RUN_CMD install $VERBOSE_ARG -Dm644 -T \
-            "${fonts}/$f" "${homeDir}/Library/Fonts/HomeManager/$f"
-        done
+        $DRY_RUN_CMD mkdir -p ${installDir}
+        $DRY_RUN_CMD ${pkgs.rsync}/bin/rsync $VERBOSE_ARG -acL --chmod=u+w --delete ${fonts}/ ${installDir}
+
+        if [[ ! -v DRY_RUN ]]; then
+          echo ${fontsEnv} > ${versionFile}
+        fi
```
Comment (berbiche) [inline modules/targets/darwin/fonts.nix:30]:
```suggestion
        $DRY_RUN_CMD echo ${escapeShellArg fontsEnv} > ${versionFile}
```
`fontsEnv` declaration could also be updated to use `escapeShellArg` instead.

File: modules/targets/darwin/fonts.nix [hunk 3]
```diff
@@ -10,18 +10,24 @@ let
...
+          return 0
+        fi
 
-        local f
-        find -L "${fonts}" -type f -printf '%P\0' | while IFS= read -rd "" f; do
-          $DRY_RUN_CMD install $VERBOSE_ARG -Dm644 -T \
-            "${fonts}/$f" "${homeDir}/Library/Fonts/HomeManager/$f"
-        done
+        $DRY_RUN_CMD mkdir -p ${installDir}
+        $DRY_RUN_CMD ${pkgs.rsync}/bin/rsync $VERBOSE_ARG -acL --chmod=u+w --delete ${fonts}/ ${installDir}
```
Comment (berbiche) [inline modules/targets/darwin/fonts.nix:26]:
```suggestion
        $DRY_RUN_CMD ${pkgs.rsync}/bin/rsync $VERBOSE_ARG -acL --chmod=u+w --delete ${escapeShellArg fonts}/ ${installDir}
```

**Review by berbiche — COMMENTED** [review]
Nice improvement 👍 
Thanks for the contribution.

**Comment by apeschar** [comment]
Thanks for your review, @berbiche. I think I've fixed your concerns.

How do you feel about merging this? :)

**Comment by apeschar** [comment]
@ncfavier, would you perhaps like to have a look at this PR?

I think I've addressed @berbiche's comments, and this will be a big performance boost for those on macOS with a lot of fonts or with large fonts (eg, Nerd Fonts). I've been using this for the past months without issues.

**Comment by ncfavier** [comment]
Do we need step 1, isn't step 2 enough?

**Comment by apeschar** [comment]
Need no, but it speeds things up significantly:

`rsync` still walks the entire tree and compares the content of all the files because it cannot rely on timestamps (since all timestamps are 0 in Nix).

So although that is faster than rewriting all the files, it still involves reading all files, which can take a few seconds.

**Comment by ncfavier** [comment]
Fair enough. Have you considered something like

```nix
home.file.${versionFile} = {
  text = fontsEnv;
  onChange = ''
    rsync ...
  '';
};
```

?

**Comment by apeschar** [comment]
I had not considered that; I was unaware of this possibility. I've implemented the same logic using onChange. Have another look, if you like.

**Review by ncfavier — APPROVED** [review]

**Comment by apeschar** [comment]
Rebased against master. I think this can safely be merged.

File: modules/targets/darwin/fonts.nix [hunk 4]
```diff
@@ -10,20 +10,17 @@ let
...
-      }
-      copyFonts
-    '';
+    home.file."Library/Fonts/.home-manager-fonts-version" = {
+      text = "${fontsEnv}";
+      onChange = ''
+        mkdir -p ${escapeShellArg installDir}
+        ${pkgs.rsync}/bin/rsync $VERBOSE_ARG -acL --chmod=u+w --delete \
+          ${escapeShellArgs ["${fonts}/" installDir]}
+      '';
```
Comment (sumnerevans) [inline modules/targets/darwin/fonts.nix:23]: These probably still need to be protected by the `run` shell function.

https://nix-community.github.io/home-manager/release-notes.xhtml#sec-release-24.05-highlights

**Review by sumnerevans — CHANGES_REQUESTED** [review]

**Comment by apeschar** [comment]
This PR is still applicable. Just rebased on master.

File: modules/targets/darwin/fonts.nix (hunk 4 above)
Comment (apeschar) [reply modules/targets/darwin/fonts.nix]: Never mind, I see that `run` is also necessary to respect the `silence` flag. I've added it. :)

**Review by sumnerevans — APPROVED** [review]

**Comment by rycee** [comment]
Thanks! Merged to master now 🙂

**Comment by andreykaipov** [comment]
Another awesome part of this PR is it fixes copyFonts on macOS for me. No idea how it ever worked for others before!

I used to get the following issue since BSD `find` doesn't have `-printf`: 

```
Activating copyFonts
find: -printf: unknown primary or operator
```

**Pull request merged**